File: csharp/Structural/12-Proxy/Proxy/JobCardAccessProxy.cs
public class JobCardAccessProxy : IJobCardAccess
{
    private readonly IJobCardAccess _registry;
    private readonly Handyman _handyman;

    private static readonly HashSet<string> _restricted =
        new(StringComparer.OrdinalIgnoreCase) { "Electrical", "Plumbing" };

    public JobCardAccessProxy(IJobCardAccess registry, Handyman handyman)
    {
        _registry = registry;
        _handyman = handyman;
    }

    public void Register(JobCard card)
    {
        Log($"REGISTER — {card}");
        _registry.Register(card);
    }

    public JobCard? GetJobCard(string jobId)
    {
        var card = _registry.GetJobCard(jobId);

        if (card is null)
        {
            Log($"GET #{jobId} — not found.");
            return null;
        }

        if (_restricted.Contains(card.Category) &&
            !_handyman.HasCertification(card.Category))
        {
            Log($"ACCESS DENIED — {_handyman.Name} lacks {card.Category} certification for Job #{jobId}.");
            throw new UnauthorizedAccessException(
                $"{_handyman.Name} is not certified for {card.Category} jobs.");
        }

        Log($"ACCESS GRANTED — {_handyman.Name} accessed {card}.");
        return card;
    }

    public IReadOnlyList<JobCard> GetAll()
    {
        Log($"GET ALL — {_handyman.Name} listed all job cards.");
        return _registry.GetAll();
    }

    private void Log(string message) =>
        Console.WriteLine($"  [{DateTime.Now:HH:mm:ss}] [Proxy] {message}");
}

File: typescript/Structural/12-Proxy/Proxy/JobCardAccessProxy.ts
import { IJobCardAccess } from "../Subject/IJobCardAccess";
import { JobCard } from "../Model/JobCard";
import { Handyman } from "../Model/Handyman";

export class JobCardAccessProxy implements IJobCardAccess {
  private static readonly RESTRICTED = new Set(["electrical", "plumbing"]);

  constructor(
    private readonly registry: IJobCardAccess,
    private readonly handyman: Handyman,
  ) {}

  register(card: JobCard): void {
    this.log(`REGISTER — ${card}`);
    this.registry.register(card);
  }

  getJobCard(jobId: string): JobCard | undefined {
    const card = this.registry.getJobCard(jobId);

    if (!card) {
      this.log(`GET #${jobId} — not found.`);
      return undefined;
    }

    if (
      JobCardAccessProxy.RESTRICTED.has(card.category.toLowerCase()) &&
      !this.handyman.hasCertification(card.category)
    ) {
      this.log(
        `ACCESS DENIED — ${this.handyman.name} lacks ${card.category} certification for Job #${jobId}.`,
      );
      throw new Error(
        `${this.handyman.name} is not certified for ${card.category} jobs.`,
      );
    }

    this.log(`ACCESS GRANTED — ${this.handyman.name} accessed ${card}.`);
    return card;
  }

  getAll(): ReadonlyArray<JobCard> {
    this.log(`GET ALL — ${this.handyman.name} listed all job cards.`);
    return this.registry.getAll();
  }

  private log(message: string): void {
    const time = new Date().toTimeString().slice(0, 8);
    console.log(`  [${time}] [Proxy] ${message}`);
  }
}
